### PR TITLE
Feature/generic request parameters

### PR DESCRIPTION
**Pull request by eran-gil**
@urielha your wish (part of it at least) has come true.
This branch adds an option to use any parameter you want when sending a request to ElasticSearch.
I removed the RoutingSource parameter as it is redundant after this change.
It also supports the (not-so) old DocumentIdSource for backwards-compatibility.

### CODE REVIEW

File: src/log4stash.Tests/Integration/ElasticSearchAppenderTests.cs [hunk 1]
```diff
@@ -38,7 +39,7 @@ public void Log_With_Custom_Id()
             string oldDocId = null;
             QueryConfiguration(appender =>
             {
-                oldDocId = appender.DocumentIdSource;
+                oldDocId = appender.RequestParameters["_id"];
```
Comment (urielha) [inline src/log4stash.Tests/Integration/ElasticSearchAppenderTests.cs]: I think this cause the tests to fail. Please use `TryGetValue()` instead

File: src/log4stash.Tests/Integration/ElasticSearchAppenderTests.cs [hunk 2]
```diff
@@ -62,8 +63,15 @@ public void Log_With_Custom_Routing()
             string oldRoutingSource = null;
             QueryConfiguration(appender =>
             {
-                oldRoutingSource = appender.RoutingSource;
-                appender.RoutingSource = "RoutingSource";
+                if (appender.RequestParameters.ContainsKey("_routing"))
```
Comment (urielha) [inline src/log4stash.Tests/Integration/ElasticSearchAppenderTests.cs]: `TryGetValue` here too

File: src/log4stash/Configuration/RequestParameterDictionary.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+
+namespace log4stash.Configuration
+{
+    public class RequestParameterDictionary : IDictionary<string, string>
```
Comment (urielha) [inline src/log4stash/Configuration/RequestParameterDictionary.cs]: In my opinion there is no need to implement `IDictionary` here.
You are just "hiding" a regular dictionary.. I think the only function which is needed here is `AddParameter`.

I saw also that you are using `RequestParameters.ToDictionary` (and I think this is the reason you had to implement  IDictionary), instead you can add another instance function to `RequestParameterDictionary` class and this function will return the desired result.

**Review by urielha — CHANGES_REQUESTED** [review]
Thanks a lot dude!
We have a little problem that the tests are failing (I think due to line 42 which I comment at)
If you won't get to it and I will find some time to invest in this feature myself, I will correct the things myself.

**Comment by urielha** [comment]
Another thing actually - I prefer to have `LogEventSmartFormatter` as the value (instead of just string)
I will try to do that.

File: src/log4stash/ElasticClient/InnerBulkOperation.cs [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
 namespace log4stash
 {
     public class InnerBulkOperation 
     {
         public string IndexName { get; set; }
         public string IndexType { get; set; }
         public object Document { get; set; }
-        public object DocumentId { get; set; }
-        public object Routing { get; set; }
+        public Dictionary<string, object> RequestParameters { get; set; }
```
Comment (urielha) [inline src/log4stash/ElasticClient/InnerBulkOperation.cs]: I think the name `IndexOperationParams` is more suitable

**Review by urielha — CHANGES_REQUESTED** [review]
This should do it for the `SmartFormatter`
Still need to update tests and in my opinion a better name for "RequestParameters" would be "IndexOperationParams"

**Comment by eran-gil** [comment]
made the rename as requested and fixed the failing tests (had to do with the LogEventSmartFormatter and a NullArgumentExcpetion).
That should do it :)

**Review by urielha — APPROVED** [review]

**Comment by urielha** [comment]
Thanks!

**Pull request merged**